routemaster/services/thread.rb

```diff
@@ -0,0 +1,62 @@
+require 'routemaster/services'
+require 'routemaster/mixins/log'
+require 'routemaster/mixins/log_exception'
+
+module Routemaster
+  module Services
+    # Abstracts out repeatedly calling `callable` on a thread.
+    # On exceptions, enqueues itself to `errq`.
+    #
+    # Long-running callables are expected to yield to check whether they may
+    # continue.
+    #
+    class Thread
+      include Mixins::Log
+      include Mixins::LogException
+
+      attr_reader :name
+      
+      def initialize(callable, name:, errq:)
+        @callable = callable
+        @name     = name
+        @errq     = errq
+        @running  = true
+        @thread   = ::Thread.new(&method(:_run))
+        @thread.abort_on_exception = true
+      end
+
+      def stop
+        _log.info { "stopping #{name}" }
+        @running = false
```

Comment:
`Thread#stop` would just interrupt the thread, without giving it a chance to clean up. It might, for instance, stop a delivery job after it delivers a batch but before it marks it as delivered (which would be fine as the data would get recovered, but would result in a redelivery)  

This pattern (with `@running`) simply signals that the thread should stop; the `wait` method below calls `Thread#join`, which waits for the thread to complete.